### ISSUE TITLE
Snapweb version 0.5.0 currently broken?

### ISSUE THREAD

**Issue by RuhanSA079** [badaix/snapweb#43]
With commit https://github.com/badaix/snapweb/commit/f861482c1b62e17970e32e7fe6c58bcf2d98ca8b
Seems like that iOS is still broken.

Aside from that, seems like we have a UI update issue, and audio.
On the other platforms (Linux), it seems like clicking on the play button starts the stream, no issues. Stopping the audio stream is also OK, but does not play the audio on subsequent clicks. Seems like the `snapstream.resume` call is the culprit?
I fixed mine by replacing the `snapstream.resume()` call with `snapstream = new SnapStream(config.baseUrl);`.

The UI play/stop button does only update when a user connects/disconnects from the webpage, and not on the 'click' action. If other users can chime in, would be great.
EDIT: It may work for some devices, but I'd rather revert to older version.

**Comment by badaix** [comment]
Thanks for testing @RuhanSA079 , I didn't have time to test this so far. The changes from PR #42  are now reverted and the upcoming snapcast v0.27 will bundle the older version. Since I was planing to release the react version in snapcast v0.28, this webkit issue might get resolved then.

**Comment by RuhanSA079** [comment]
As for the UI, wouldn't it be safer to call `show()` after the play button has been clicked?
Like, in two places ( after `play()` and after `stop()`)?
or once after the `playElem.click` code calls?

**Comment by bendschs** [comment]
same for me, tried with the latest development build of snapcast and snapweb, still no audio in safari.

**Comment by RuhanSA079** [comment]
@bendschs I tried something last night, and it appeared to have worked on my iPhone 7 running iOS 15.7.3.
Try applying this fix from another member: https://github.com/badaix/snapweb/compare/webkit...bakaiadam:snapweb:webkit
So, latest 0.5.0 version (not the reverted one earlier, must have the `isPlaying()` function) with this aplpied.
Also make sure to stop your snapserver, apply the compiled changes, start the snapserver.
I did have to add a additional `show()` call just inside the `playElem.click()` event, to update the UI.

**Comment by bendschs** [comment]
hi Ruhan

i get a typscript error when i try compiling this.

```
snapweb $ make
tsc --build page/tsconfig.json
page/snapcontrol.ts:324:57 - error TS2339: Property 'getClientId' does not exist on type 'typeof SnapStream'.

324             props = this.getStreamFromClient(SnapStream.getClientId()).properties;
                                                            ~~~~~~~~~~~

page/snapcontrol.ts:325:60 - error TS2339: Property 'getClientId' does not exist on type 'typeof SnapStream'.

325             metadata = this.getStreamFromClient(SnapStream.getClientId()).properties.metadata;
                                                               ~~~~~~~~~~~

page/snapcontrol.ts:505:67 - error TS2339: Property 'getClientId' does not exist on type 'typeof SnapStream'.

505             let group: Group = this.getGroupFromClient(SnapStream.getClientId());
                                                                      ~~~~~~~~~~~


Found 3 errors.

make: *** [Makefile:6: tsc] Fehler 1
```

**Comment by RuhanSA079** [comment]
@bendschs Are you on the latest version of Typescript?

**Comment by bendschs** [comment]
```
/usr/share/snapserver/snapweb $ tsc
Version 4.9.5
tsc: The TypeScript Compiler - Version 4.9.5 
```

**Comment by RuhanSA079** [comment]
That's funny why the TSC don't want to compile it. Let me redownload the source and try again (I made some other changes to mine, UI-wise)

**Comment by RuhanSA079** [comment]
@bendschs Try cloning from my new fork, I added the changes that I tested with.
the TSC on my computer compiles it fine. (Version 4.9.4)
`~/# git clone https://github.com/RuhanSA079/snapweb --depth=1`
then cd into the snapweb directory.
`~/# cd snapweb`
Make it normally.
`~snapweb/# make`
Should give you a list of "copied" assets.
I will later share the compiled JS files.

**Comment by bendschs** [comment]
thank you! i can confirm this works in general, i have slightly stuttering sound from time to time (only in safari) but i guess that could be fixed/optimised somehow.

**Comment by RuhanSA079** [comment]
OK, cool. 
So probably will make a PR at another time, or wait for the React update?

**Comment by RuhanSA079** [comment]
For the other users, if you can, please test this branch:
This changes is NOT mine, but rather just merged a few changes from other user's feedback.
This branch may or may not fix iOS audio issues, YMMV. (iOS 15+ seems to suspend the audio in the web browser, does not allow it to continue with even clicking the Play button in the Control Center)
https://github.com/RuhanSA079/snapweb

**Comment by badaix** [comment]
Please also give the react version a try: https://github.com/badaix/snapweb/tree/react

![image](https://user-images.githubusercontent.com/10322999/215767140-0ef2bfb2-ff4e-4103-b976-5bf6ac1c587e.png)

There are prebuilt versions available in github actions: 
https://github.com/badaix/snapweb/actions/runs/4047477321

The project itself comes with a docker container that contains the react development tools and can be used transparently with VSCode. This react version will be shipped with the next Snapcast release and hence all fixes made now to the develop/master branch must be ported also to the react branch. Please note that I hardcoded my server's IP in index.tsx. This must be changed of course to localhost or whatever server you want to use.

**Comment by badaix** [comment]
> OK, cool. So probably will make a PR at another time, or wait for the React update?

Yes, please wait for the react update. I'm using it since a few month and like it much more than the current versions.

**Comment by bendschs** [comment]
getting typscript errors again, while trying to build the react version.

looks very nice btw!


```
tsc --build page/tsconfig.json
page/snapstream.ts:671:24 - error TS2304: Cannot find name 'Flac'.

671         this.decoder = Flac.create_libflac_decoder(true);
                           ~~~~

page/snapstream.ts:673:31 - error TS2304: Cannot find name 'Flac'.

673             let init_status = Flac.init_decoder_stream(this.decoder, this.read_callback_fn.bind(this), this.write_callback_fn.bind(this), this.error_callback_fn.bind(this), this.metadata_callback_fn.bind(this), false);
                                  ~~~~

page/snapstream.ts:675:13 - error TS2304: Cannot find name 'Flac'.

675             Flac.setOptions(this.decoder, { analyseSubframes: true, analyseResiduals: true });
                ~~~~

page/snapstream.ts:693:45 - error TS2304: Cannot find name 'Flac'.

693         while (this.flacChunk.byteLength && Flac.FLAC__stream_decoder_process_single(this.decoder)) {
                                                ~~~~

page/snapstream.ts:694:13 - error TS2304: Cannot find name 'Flac'.

694             Flac.FLAC__stream_decoder_get_state(this.decoder);
                ~~~~

page/snapstream.ts:707:43 - error TS2503: Cannot find namespace 'Flac'.

707     read_callback_fn(bufferSize: number): Flac.ReadResult | Flac.CompletedReadResult {
                                              ~~~~

page/snapstream.ts:707:61 - error TS2503: Cannot find namespace 'Flac'.

707     read_callback_fn(bufferSize: number): Flac.ReadResult | Flac.CompletedReadResult {
                                                                ~~~~

page/snapstream.ts:725:59 - error TS2503: Cannot find namespace 'Flac'.

725     write_callback_fn(data: Array<Uint8Array>, frameInfo: Flac.BlockMetadata) {
                                                              ~~~~

page/snapstream.ts:761:9 - error TS2304: Cannot find name 'Flac'.

761         Flac.FLAC__stream_decoder_process_until_end_of_metadata(this.decoder);
            ~~~~


Found 9 errors.
```

**Comment by bendschs** [comment]
I tried to use Visual Studio Code, but that seems a bit beyond my knowledge ...

**Comment by RuhanSA079** [comment]
@bendschs 
You will probably have to install multipass to spin up a VM to avoid contamination of your build-env, or you could run this on your Linux machine:

Install node v16.x by doing this in your shell: (multipass highly recommended)
`curl -s https://deb.nodesource.com/setup_16.x | sudo bash`
After install, install nodejs 16.x:
`sudo apt install nodejs`
If you had an older node install, remove it by `sudo apt remove nodejs` and then if libnode72 bothers you, remove that lib as well.
Check your node version, mine is 16.19.0: `node -v`
Clone the repo: `git clone https://github.com/badaix/snapweb --branch=react --depth=1`
cd into the repo: `cd snapweb`
Clean the npm cache: `npm cache clean --force`
Install components: `npm ci`
Run npm install: `npm install`
After that, set one extra environment variable: `export GENERATE_SOURCEMAP=false`
After that, you build-env should be ready to build the react app, so go make a sandwich or have a cuppa while it builds.
top-tip: Please do give the VM some moar CPU and memory, if your PC/laptop can allow it (npm will chow CPU-time and some RAM for lunch)
I gave mine two CPUs, and 4GB of RAM. (Running on a i3-6006U lappy, it aint much, but it works.)
To build: `npm run build`
The output should be in the "build" folder.
I will now update my build index.tsx and App.tsx file to refer to my server's address, and rebuild with `npm run build` again, to test the audio stream stuff.
There's two places needed to reflect your server's IP and portno.
Since multipass is inside a container, you would need to use `multipass transfer` to transfer the files out, or I zipped them up in a zip file, done the transfer once.

**Comment by bendschs** [comment]
thanx Ruhan for your accurate instructions, i was able to test the react version. in general it works but regarding safari/ios it show the same behaviour as the last official release. also i did not get the meta data to be shown in the snapweb interface (using it together with mopidy.

**Comment by RuhanSA079** [comment]
> thanx Ruhan for your accurate instructions, i was able to test the react version. in general it works but regarding safari/ios it show the same behaviour as the last official release. also i did not get the meta data to be shown in the snapweb interface (using it together with mopidy).

Glad to hear that you were able to set it up and test.
Unfortunately, I ran into some other issues, my plan was to use the snapcast system with audio input to serve up some webclients, strip out the controls, leaving only the play/stop button on the webpage. The general idea was to use this as a live-audio translation system, but due to iOS issues, I found a better purpose-built piece of software, that does the audio over WebRTC for both Android and iOS. 

I will keep an eye on this system/webpage, controlling my own speakers in each room throughout the house some day. I wish you all the best of luck.
I think I can close this thread, as the iOS timing issues with Webkit is known, and stutters sometimes and in some cases, no audio.

There's maybe an idea to pipe the audio to the clients over WebRTC, at least for those who operates the webpage on some device, and leave the protocol to the snapcast clients.

At-least I was able to test the other users' changes on my fork.
Closing thread.
